# Artificial Neural Network Practice

The goal is to predict if customer is going to leave a bank using ANN. The data set contains information about credit score, country, gender, age of customer, etc. The predicted value is binary. The data set was taken from course.

## Importing the libraries

In [4]:
import numpy as np
import pandas as pd
import tensorflow as tf

## Data Preprocessing

In [6]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

### Encoding categorical data

Label Encoding for the "Gender" column and One Hot Encoding for the "Geography" column

In [7]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])

In [8]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))

### Splitting the dataset into the Training set and Test set

In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

In [10]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Building the ANN

Initializing, adding two hidden layer (input is added automatically) and output layer

In [11]:
ann = tf.keras.models.Sequential()
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Compiling the ANN

In [13]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

## Training the ANN

In [14]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 1s 2ms/step - loss: 0.6451 - accuracy: 0.6279: 0s - loss: 0.7014 - accura
Epoch 2/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4927 - accuracy: 0.7950
Epoch 3/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4524 - accuracy: 0.7962
Epoch 4/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4377 - accuracy: 0.7971
Epoch 5/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4278 - accuracy: 0.8008: 0s - loss: 0.4224 - 
Epoch 6/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4171 - accuracy: 0.8105
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4025 - accuracy: 0.8245
Epoch 8/100
250/250 [==============================] - 0s 2ms/step - loss: 0.3853 - accuracy: 0.8316
Epoch 9/100
250/250 [==============================] - 0s 2ms/step - loss: 0.3701 - accuracy: 0.8439
Epoch 10/100
250/250 [===================

## Making the predictions and evaluating the model

Predicting for customer: Geography: France. Credit Score: 600. Gender: Male. Age: 40 years old. Tenure: 3 years. Balance:  60000.
Number of Products: 2. Does this customer have a credit card ? Yes. Is this customer an Active Member: Yes. Estimated Salary: 50000

In [15]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

[[False]]


ANN model predicts that this customer stays in the ban

### Predicting the Test set results

In [16]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


## Making the Confusion Matrix

In [17]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1510   85]
 [ 198  207]]


0.8585

So it makes 198 FN and 85 FP predictions, but 85% is good result